src/instructlab/model/download.py

```diff
@@ -1,31 +1,258 @@
 # SPDX-License-Identifier: Apache-2.0
 
 # Standard
+from pathlib import Path
+import abc
+import logging
 import os
+import re
+import subprocess
 
 # Third Party
 from huggingface_hub import hf_hub_download, list_repo_files
 from huggingface_hub import logging as hf_logging
 from huggingface_hub import snapshot_download
+from packaging import version
 import click
 
 # First Party
 from instructlab import clickext
 from instructlab.configuration import DEFAULTS
+from instructlab.utils import _extract_SHA, _load_json, is_huggingface_repo, is_oci_repo
+
+logger = logging.getLogger(__name__)
+
+
+class Downloader(abc.ABC):
+    """Base class for a downloading backend"""
+
+    def __init__(
+        self,
+        repository: str,
+        release: str,
+        download_dest: str,
+    ) -> None:
+        self.repository = repository
+        self.release = release
+        self.download_dest = download_dest
+
+    @abc.abstractmethod
+    def download(self) -> None:
+        """Downloads model from specified repo/release and stores it into download_dest"""
+
+
+class HFDownloader(Downloader):
+    """Class to handle downloading safetensors and GGUF models from Hugging Face"""
+
+    def __init__(
+        self,
+        repository: str,
+        release: str,
+        download_dest: str,
+        filename: str,
+        hf_token: str,
+        ctx,
+    ) -> None:
+        super().__init__(
+            repository=repository, release=release, download_dest=download_dest
+        )
+        self.filename = filename
+        self.hf_token = hf_token
+        self.ctx = ctx
```

Comment:
nit: both implementations receive `ctx`. You could pass it to base class too and set it there to `self.ctx`.